### ISSUE TITLE
Is it possible to start mintty on Linux with wine64?

### ISSUE THREAD

**Issue by csdvrx** [mintty/mintty#1096]
To work on the touchscreen and pen stylus support for #1073 , I am trying to use wine64 to run mintty - given what I read on https://github.com/msys2/MSYS2-packages/issues/682 this should work in 2021 with wine64 set as Windows8.1 intead of XP.

However, when trying to start msys2 installation, I get a lackadaisical read access error on 00000001405BB3D0

Is there a standalone mintty cygwin binary package available, or instructions to compile mintty on Ubuntu 20 LTS with the goal of running mintty under wine64? (version 5.0-3ubuntu1 is installed by default)

/usr/lib/wine/wine64 ~/Downloads/msys2-x86_64-20210228.exe
0009:err:winediag:SECUR32_initNTLMSP ntlm_auth was not found or is outdated. Make sure that ntlm_auth >= 3.0.25 is in your path. Usually, you can find it in the winbind package of your distribution.
0009:fixme:console:GetConsoleProcessList (0x32f990,2): stub
0009:fixme:win:RegisterPowerSettingNotification (0x10054,{02731015-4510-4526-99e6-e5a17ebd1aea},0): stub
wine: Unhandled page fault on read access to 0000000000000001 at address 00000001405BB3D0 (thread 0009), starting debugger...

Any suggestion?

**Comment by Biswa96** [comment]
Wine can **not** run any cygwin programs. See the bug list here https://bugs.winehq.org/buglist.cgi?quicksearch=cygwin

**Comment by csdvrx** [comment]
Is there any possibility to run mintty on Linux, like by compiling it differently so that wine can use the binary?

**Comment by Biswa96** [comment]
Hmm... Interesting. I have not thought about. There are hundreds or thousands terminal emulator programs in any GNU/Linux distribution. If cygwin will work with wine we may not even need montry.

**Comment by Biswa96** [comment]
Hmm... Interesting. I have not thought about. There are hundreds or thousands terminal emulator programs in any GNU/Linux distribution. If cygwin will work with wine we may not even need mintty.

**Comment by mintty** [comment]
Actually I can start mintty with wine, it's just not able to run any child program (like a shell, cmd.exe).
Wine seems to be quite an unmature project even after many years, so I have no idea how or when it might be possible to tweak mintty to work with wine.
Another idea might be to port mintty directly to Linux (if that meets your use case) but unfortunately, the X11 backend once available in putty is not integrated anymore, so that would be another lot of effort.

**Comment by avih** [comment]
> the X11 backend once available in putty

I believe putty linux support is still maintained to this day. I think the config is GTK based, I don't know if the terminal itself can be built as pure X11 without GTK support.

**Comment by mintty** [comment]
@avih, I've thought about back-porting to X11 already, or maybe to Wayland if that's easier. There are two problems though:
- Mintty makes use of the Windows API for a lot of new features since the spawn-off from putty, it will not be sufficient to re-integrate the previous backend.
- There is no X11 API know-how on my side.

**Comment by avih** [comment]
Personally I don't think porting mintty to X11 or any other *nix backend is worth the effort and potential increased complexity. Mintty is great, but there are plenty of terminals for unix a;ready, including putty (I think I now recall that it has a hard dependency on GTK).

**Comment by BrianInglis** [comment]
It's also available as a text term under the name putty-nogtk from some distros

**Comment by avih** [comment]
> It's also available as a text term under the name putty-nogtk from some distros

Yes, but I don't think it includes the terminal (it does have puttygen and few more). For instance, this is the FreeBSD amd64 package: https://pkg.freebsd.org/FreeBSD:13:amd64/latest/All/putty-nogtk-0.74_1.txz

Long time ago I tried to enhance its terminal key handling to accept more combinations (ctrl+arrows etc), and the key handling code was definitely GTK. Maybe it also has non-GTK terminal/frontend, but I don't recall seeing it back then.

**Comment by csdvrx** [comment]
>  Mintty is great, but there are plenty of terminals for unix already, 

@avih : There are plenty on terminals on Linux that each suck *VERY* badly if you try to do anything more complicated than running a shell.

The general quality of terminals on Linux is so bad that it's not even funny. xterm is the least worst, after some heavy configuration (cf github.com/csdvrx/cutexterm  ) but it's missing many features that mintty has - on top of my head, programming ligatures and resizing of sixel images when changing fontsize. 

I would prefer to run the same terminal on Windows and Linux - and no, before someone mentions mlterm, I know it exists, but it has its own set of issues.

Why settle on bad terminals? mintty is great. Let's make it work!

> Wine seems to be quite an unmature project even after many years,

@mintty  : I use Excel and a few other Window-only applications everyday thanks to wine. It works great!

If wine can supports Office and various games, I think it should be able to support mintty - especially since potential issues could be fixed in mintty source, which is not possible with say Excel.

If compiling mintty under cygwin to make it work on Linux is possible (even after patching the toolchain if necessary!) I'll be happy to do that and keep a repo to provide packages to others also looking for a good terminal experience on Linux

**Comment by mintty** [comment]
Speaking about Linux terminals I am especially startled by the issue that many of them need a minute or more to even start up.
Considering that, the only ones I can take serious are xterm, urxvt, vte, st. Which makes me agree that porting mintty to Linux would be a nice endeavour.

As commented above, mintty does in fact run under wine, but due to some cygwin forking issues it's not useful as I couldn't get it to spawn any client program.

However, if this is going to be a project at all, I'd rather favour a native port than fiddling with wine issues. And a roadmap that comes to my mind is to port to GTK rather than X/Wayland directly, and develop that in cygwin as long as Windows API calls are also needed. Problem again is that I have no personal GTK know-how either.

By the way, I cannot promise that such a port would support programming ligatures as those are support by Windows GDI font rendering implicitly, so they'd need be supported by GTK/X11 then.

**Comment by avih** [comment]
> many of them need a minute or more to even start up

Surely you mean a second? But personally I've not seen a terminal, on any OS, which takes much less than 100ms to launch...

**Comment by mintty** [comment]
> Surely you mean a second?

No, I meant in fact a minute. Well, that was on one system I had, an enterprise PC with an aggressive virus scanning policy.
I guess the fact that gnome-terminal and many of the other desktop-system related ones have more than 60 non-Windows DLL dependencies may have caused this issue.

**Comment by avih** [comment]
> No, I meant in fact a minute. ... aggressive virus scanning ... gnome-terminal ... more than 60 non-Windows DLL dependencies

To me this sounds like a pathological bad case of enterprise + cygwin + gnome + dbus slowness. It's a really bad combo of factors.

On my cygwin setup, st and/or xterm launch pretty much instantly. I imagine urxvt would be similar. I didn't test VTE-based terminals in cygwin.

However, if you test on any non-windows system - and this thread is about using mintty and its advantages on linux - then I doubt you can find a terminal which takes even 100ms to launch.

**Comment by BrianInglis** [comment]
I seem what you mean with lxterminal > 50 DLLs compared to mintty > 30. 
However the (silent) profile scripts seem to run quickly (comparable to WSL or a VM) under lxterminal compared to mintty. 
Normally it's the profile scripts that take the time going from window display to prompt - mintty Alt-F2 takes < 1s - Cygwin profile scripts take many seconds, compared to WSL which takes a few, or a VM term which takes < 1s, for ~1500 profile lines, ~700 bash completions and about ~400 Cygwin specific, compared to only ~70 Linux with minimal bash completions.

**Comment by mintty** [comment]
I was speaking about the time until just the window appears... and mintty has only 1 non-Windows dependency (the cygwin DLL).
I guess that virus scanners might be faster on Windows DLLs perhaps (just speculating) or not scan them at all.

**Comment by WSLUser** [comment]
I would make a comparison to Kitty for performance and possibly even other cross-platform terminals like Contour to gauge what it would take for Mintty to become available on more platforms. I know Contour at least already has a GTK backend and attempts to do native rendering though hopefully someone will come along and add Vulkan rendering as it's more performant with higher availability for different platforms and archetectures (I believe they use native APIs when needed to be similar to DirectWrite/Metal,etc.).  

> I guess that virus scanners might be faster on Windows DLLs perhaps (just speculating) or not scan them at all.

This is still mostly the case though there are more virus scanners supporting Linux now thanks to WSL among other factors. Still pretty slow for even Windows dlls for most virus scanners.
